test/apps/iodemo/run_io_demo.sh

```diff
@@ -0,0 +1,539 @@
+	do
+		case ${map_by} in
+		node)
+			# For each node, add both servers and clients, the last node may
+			# have less than others
+			node_num_servers=$(min ${max_servers_per_node} $((num_servers - added_servers)))
+			node_num_clients=$(min ${max_clients_per_node} $((num_clients - added_clients)))
+			;;
+		slot)
+			# Servers take slots first, and clients take what is left
+			node_num_servers=$(min ${tasks_per_node} $((num_servers - added_servers)))
+			node_num_clients=$(min $((tasks_per_node - node_num_servers)) \
+			                       $((num_clients - added_clients)))
+			;;
+		esac
+
+		num_servers_per_host[${host}]=${node_num_servers}
+		num_clients_per_host[${host}]=${node_num_clients}
+
+		for ((i=0;i<node_num_servers;++i))
+		do
+			port_num=$((base_port_num + i))
+			client_connect_list+=" ${ip_address_per_host[${host}]}:${port_num}"
+		done
+
+		added_servers=$((added_servers + node_num_servers))
+		added_clients=$((added_clients + node_num_clients))
+	done
+
+	#
+	# In verbose mode, make following changes:
+	#   1. show mixed output in stdout (in addition to log files)
+	#   2. run the spawned scripts with '-x'
+	#   3. show output from kill commands
+	#
+	if is_verbose
+	then
+		show_var client_connect_list
+		sh_args="-x"
+		wait_redirect=""
+		log_redirect="|& tee "
+	else
+		sh_args=""
+		wait_redirect=">& /dev/null"
+		log_redirect=">& "
+	fi
+
+	exe_basename=$(basename ${iodemo_exe})
+
+	# create run scripts
+	for host in $(split_list ${host_list})
+	do
+		command_file="iodemo_commands_${host}.sh"
+
+		# Add file header and startup
+		cat >${command_file} <<-EOF
+			#!/bin/sh ${sh_args}
+			#
+			# Launch script for io_demo on ${host} with ${num_servers_per_host[${host}]} servers and ${num_clients_per_host[${host}]} clients
+			#
+
+			kill_iodemo() {
+			    # kill process by exact name to avoid killing launch scirpts
+			    # lt- prefix is in case iodemo is wrapped by libtool
+			    for pattern in ${exe_basename} lt-${exe_basename}
+			    do
+			        pkill --exact -u ${USER} -9 \${pattern}
+			    done
+			}
+
+			signal_handler() {
+			    echo "Got signal, killing all iodemo processes"
+			    kill_iodemo
+			}
+
+			# kill existing processes and trap signals
+			trap signal_handler INT TERM
+			kill_iodemo
+
+			# Use same work dir as launch script
+			cd $PWD
+
+			# UCX inherited environment variables
+			EOF
+
+		# Add all relevant env vars which start with UCX_ to the command file
+		env | grep -P '^UCX_.*=|^PATH=|^LD_LIBRARY_PATH=' | \
+			xargs -L 1 echo export >>${command_file}
+
+		# Add servers commands
+		cat >>${command_file} <<-EOF
+
+			# Server commands
+			EOF
+
+		cmd_prefix="stdbuf -e0 -o0 timeout -s 9 $((duration + client_wait_time))s"
+		for ((i=0;i<${num_servers_per_host[${host}]};++i))
+		do
+			port_num=$((base_port_num + i))
+			log_file=$(printf "iodemo_%s_server_%02d.log" ${host} $i)
+			echo ${log_file}
+			cat >>${command_file} <<-EOF
+				${cmd_prefix} \\
```

Comment:
also, every server and client command to start - should be generated as separate function, i.e.
```
start_client_01() {
}
start_client_02() {
}
```
and generated command script should also call them as function.
Once we have it - we can add command line to that generated script to start specific function:

```
%iodemo_commands_mtr-vdi-326.sh --start client_01
```
it will dynamically call function start_client_01

WDYT

Reply:
as below (partially impletemented)

``` bash
diff --git a/test/apps/iodemo/run_io_demo.sh b/test/apps/iodemo/run_io_demo.sh
index 59398530..fa024916 100755
--- a/test/apps/iodemo/run_io_demo.sh
+++ b/test/apps/iodemo/run_io_demo.sh
@@ -452,12 +452,15 @@ make_scripts()
                do
                        port_num=$((base_port_num + i))
                        log_file=$(printf "iodemo_%s_server_%02d.log" ${host} $i)
+                       cmd_prefix+=" env IODEMO_ROLE=server-${i} "
                        echo ${log_file}
                        cat >>${command_file} <<-EOF
+                       function start_server_${i}() {
                                ${cmd_prefix} \\
                                    ${iodemo_exe} \\
                                        ${iodemo_server_args} -p ${port_num} \\
                                        ${log_redirect} ${log_file} &
+                       }
                                EOF
                done
 
@@ -478,12 +481,15 @@ make_scripts()
                for ((i=0;i<num_clients_per_host[${host}];++i))
                do
                        log_file=$(printf "iodemo_%s_client_%02d.log" ${host} $i)
+                       cmd_prefix+=" env IODEMO_ROLE=client-${i} "
                        echo ${log_file}
                        cat >>${command_file} <<-EOF
+                       function start_client_${i}() {
                                ${cmd_prefix} \\
                                    ${iodemo_exe} \\
                                        ${iodemo_client_args} ${client_connect_list} \\
                                        ${log_redirect} ${log_file} &
+                       }
                                EOF
                done

```

Reply:
> it will dynamically call function start_client_01
> 
> WDYT

IMO it should be in next PR since the way the MTT will stop/start processes still needs to be decided

Reply:
ok